src/transformers/pipelines/fill_mask.py

```diff
@@ -157,32 +151,40 @@ def __call__(self, *args, targets=None, top_k: Optional[int] = None, **kwargs):
                     values = tf.gather_nd(values, tf.reshape(sort_inds, (-1, 1))).numpy()
                     predictions = target_inds[sort_inds.numpy()]
             else:
-                masked_index = torch.nonzero(input_ids == self.tokenizer.mask_token_id, as_tuple=False)
+                masked_indices = torch.nonzero(input_ids == self.tokenizer.mask_token_id, as_tuple=False)
+                # Fill mask pipeline supports at least one ${mask_token} per sample
+                self.ensure_atleast_one_mask_token(masked_indices.numpy())
 
-                # Fill mask pipeline supports only one ${mask_token} per sample
-                self.ensure_exactly_one_mask_token(masked_index.numpy())
+                logits_multiple = [outputs[i, index.item(), :] for index in masked_indices]
+
+                probs_multiple = [logits.softmax(dim=0) for logits in logits_multiple]
 
-                logits = outputs[i, masked_index.item(), :]
-                probs = logits.softmax(dim=0)
                 if targets is None:
-                    values, predictions = probs.topk(top_k if top_k is not None else self.top_k)
+                    values_all = []
+                    predictions_all = []
+                    for probs in probs_multiple:
+                        values, predictions = probs.topk(top_k if top_k is not None else self.top_k)
```

Comment:
This is debattable.

Are the proposition single tokens for mask tokens, or are they tuples of answers. Consider the following:

`This <mask> is to <mask> what rice is to sushi.`

Here are the top-3 proposition for the 3 masks:
[apple, rhubarb, Paris]
[pie, France, biking]

With your code, you are going to propose IIUC
(apple, pie)
(rhubarb, France)
(Paris, biking)

It's possible (not necessarely though) that the propositions you want to make are more like:

(apple, pie)
(rhubarb, pie)
(Paris, France).

My suggestion at this point it to look at joint probabilities for the tuple suggestion instead of token per token.

Reply:
@Narsil  This is correct, I have been a little worried about how BERT's masking for multiple masks and how one obtains the joint prob instead of a single token specific probability. Since it is simultaenously making the prediction for all the masks, it tends to make more mistakes(both gramatically) and knowledge wise too. I would be grateful if you could help me understand how one retrieves a join probability in this case.

This issue gets worse  when the masks are situated closer to each other, with BERT typically predicting the same word for both the mask slots.

Reply:
I don't think you can do correct joint probabilities.
The output is **by design** the sum of all joint probabilities at every locus.
What I meant is Bert *cannot* outptut {token1:"either", token2:"or", score:50%}, {token1:"or", token2:"either", score:50%}. It *has* to output {token1: ["either", 50%], ["or", 50%]} {token2: ["either", 50%], ["or", 50%]}. So you have no way of recovering the first proposed solution and your best guess can only be (either either, 25), (either, or, 25), (or, either, 25), (or, or, 25) 

What I was suggesting, as a better guess, was simply treating them like they were:

- Softmax all mask locus independantly
- create all joint probabilities (lazily because it's a combinatorial)
  - p1_1 x p2_1
  - p1_1 x p2_2
  - ...
  - p1_2 x p2_1
  - ....
  - px_y where x is the location of the max token, and y is the rank of the proposed token
- Softmax that joint probabilities list (just so that output scores are correctly scaled, could be ignored because of combinatorial)
- Order them

Reply:
This makes sense to me, awesome, I will get on this.

Reply:
This will rank ('apple', 'France') and ('Paris', 'pie') higher than ('Paris', 'France'). We need some measure how happy the transformer is with each candidate sequence. I think we need additional forward passes to measure the effect of each combination. If there is some way of measuring the model's happiness with a candidate sequence one pass per candidate sequence will suffice. If not, I'd suggest to run
```
This apple is to <mask> what rice is to sushi.
This rhubarb is to <mask> what rice is to sushi.
This Paris is to <mask> what rice is to sushi.
This <mask> is to pie what rice is to sushi.
This <mask> is to France what rice is to sushi.
This <mask> is to biking what rice is to sushi.
```
and then multiple the probabilities. We will need some kind of beam search to limit the combinations tested as the number of forward passes needed will otherwise explode for more masked tokens or large `top_k`.